source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -224,6 +227,9 @@ void AuthenticatorImpl::verifyKey() {
     // Remove JWT from headers.
     curr_token_->removeJwt(*headers_);
   }
+  if (set_payload_cb_ && provider.payload_in_metadata()) {
+    set_payload_cb_(provider.issuer(), jwt_->payload_str_base64url_);
```

Comment:
Actually can we parse the JSON payload into a ProtobufWkt::Struct? otherwise you won't be able to match each claim in RBAC.

Also rather than issuer as key, provider_name would be better.

Reply:
For allow_missing_or_fail rule,  it doesn't need to specify provider,  so there is not provider_name.  In theory,  there could be two ProviderInfo with same same issuer.  It is a useful feature for normal cases.  But for allow_missing_or_fail case, we choose the first one matched with the issuer from the JWT token,  it is better not to have two providers points to the same issuer in the config.
Overall,  writing issuer works for all cases, but not provider name.